rpcs3/Emu/Io/RB3MidiGuitar.cpp

```diff
@@ -0,0 +1,307 @@
+// Rock Band 3 MIDI Pro Adapter Emulator (Guitar Mode)
+
+#include "stdafx.h"
+#include "RB3MidiGuitar.h"
+#include "Emu/Cell/lv2/sys_usbd.h"
+
+LOG_CHANNEL(rb3_midi_guitar_log);
+
+usb_device_rb3_midi_guitar::usb_device_rb3_midi_guitar(const std::array<u8, 7>& location, const std::string& device_name, bool twentytwo_fret)
+	: usb_device_emulated(location)
+{
+	// For the 22-fret guitar (Fender Squier), the only thing that's different
+	// is the device ID reported by the MIDI Pro Adapter.
+	//
+	// Everything else is *exactly* the same as the 17-fret guitar (Fender Mustang).
+	if (twentytwo_fret)
+	{
+		device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x0200, 0x00, 0x00, 0x00, 64, 0x12ba, 0x2538, 0x01, 0x01, 0x02, 0x00, 0x01});
+	}
+	else
+	{
+		device = UsbDescriptorNode(USB_DESCRIPTOR_DEVICE, UsbDeviceDescriptor{0x0200, 0x00, 0x00, 0x00, 64, 0x12ba, 0x2438, 0x01, 0x01, 0x02, 0x00, 0x01});
+	}
+
+	auto& config0 = device.add_node(UsbDescriptorNode(USB_DESCRIPTOR_CONFIG, UsbDeviceConfiguration{41, 1, 1, 0, 0x80, 32}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_INTERFACE, UsbDeviceInterface{0, 0, 2, 3, 0, 0, 0}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_HID, UsbDeviceHID{0x0111, 0x00, 0x01, 0x22, 137}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x81, 0x03, 0x0040, 10}));
+	config0.add_node(UsbDescriptorNode(USB_DESCRIPTOR_ENDPOINT, UsbDeviceEndpoint{0x02, 0x03, 0x0040, 10}));
+
+	usb_device_emulated::add_string("Licensed by Sony Computer Entertainment America");
+	usb_device_emulated::add_string("Harmonix RB3 MIDI Guitar Interface for PlayStation®3");
+
+	// connect to midi device
+	midi_in = rtmidi_in_create_default();
+	ensure(midi_in);
+
+	rb3_midi_guitar_log.notice("Using %s API", rtmidi_api_name(rtmidi_in_get_current_api(midi_in)));
+
+	if (!midi_in->ok)
+	{
+		rb3_midi_guitar_log.error("Could not get MIDI in ptr: %s", midi_in->msg);
+		return;
+	}
+
+	rtmidi_in_ignore_types(midi_in, false, true, true);
+
+	const s32 port_count = rtmidi_get_port_count(midi_in);
+
+	if (port_count == -1)
+	{
+		rb3_midi_guitar_log.error("Could not get MIDI port count: %s", midi_in->msg);
+		return;
+	}
+
+	for (s32 port_number = 0; port_number < port_count; port_number++)
+	{
+		char buf[128]{};
+		s32 size = sizeof(buf);
+		if (rtmidi_get_port_name(midi_in, port_number, buf, &size) == -1)
+		{
+			rb3_midi_guitar_log.error("Error getting port name for port %s:", port_number, midi_in->msg);
```

Comment:
before you merge this needs to be fixed

Reply:
then please rebase and fix in a new commit